File: test/c++/anderson.cpp
#include <triqs_ctint/solver_core.hpp>

#include <h5/h5.hpp>
#include <triqs/gfs.hpp>
#include <triqs/mesh.hpp>
#include <triqs/test_tools/gfs.hpp>

using namespace triqs_ctint;

int main(int argc, char *argv[]) {
  mpi::environment env(argc, argv);

  // System Parameters
  double U  = 2.0;
  double mu = U / 2;

  // Discrete bath energies and hoppings
  std::vector<double> energ = {-0.7, -0.15, 0.15, 0.7}; // Vertex Paper ( optimized for Beta=20 ) to fit box DOS
  std::vector<double> hopp  = {0.45, 0.34, 0.34, 0.45};

  // The alpha function
  alpha_t alpha(2);
  double delta = 0.35;
  alpha[0]     = {{0.5 + delta, 0.5 - delta}};
  alpha[1]     = {{0.5 - delta, 0.5 + delta}};

  // Construct Parameters
  constr_params_t pc;
  pc.beta      = 20.0;
  pc.gf_struct = {{"up", 1}, {"down", 1}};
  pc.n_tau     = 10000;
  pc.n_iw      = 500;
  pc.use_D     = false;
  pc.use_Jperp = false;

  // Solve Parameters
  solve_params_t ps;
  ps.h_int                = U * n("up", 0) * n("down", 0);
  ps.n_s                  = 2;
  ps.alpha                = alpha;
  ps.n_cycles             = 1000000;
  ps.length_cycle         = 10;
  ps.n_warmup_cycles      = 10000;
  ps.use_double_insertion = false;
  ps.max_time             = -1;
  ps.post_process         = true;
  ps.n_cheb_coeffs        = 50;

  ps.measure_histogram = true;

  ps.measure_M_tau      = false;
  ps.measure_M_tau_cheb = true;
  ps.measure_M_iw       = false;

  //ps.measure_M4_iw  = true;
  //ps.n_iw_M4        = 32;

  //ps.measure_M3pp_iw  = true;
  //ps.measure_M3ph_iw  = true;
  //ps.n_iw_M3          = 64;

  //ps.measure_M3pp_tau  = true;
  //ps.measure_M3ph_tau  = true;
  //ps.n_tau_M3           = 1000;

  //ps.measure_chi2pp_tau  = true;
  //ps.measure_chi2ph_tau  = true;
  //ps.n_tau_chi2          = 1000;
  //ps.n_iw_chi2           = 128;

  ps.nfft_buf_size = 50;

  solver_core S(pc);

  // set up hybridization delta(i\omega_n)
  auto delta_w = gf<imfreq, matrix_valued>{{pc.beta, Fermion, pc.n_iw}, make_shape(1, 1)};
  delta_w()    = 0;
  for (auto i : range(energ.size())) {
    auto term = gf<imfreq, matrix_valued>{{pc.beta, Fermion, pc.n_iw}, make_shape(1, 1)};
    term(iw_) << hopp[i] * hopp[i] / (iw_ - energ[i]);
    delta_w += term;
  }

  // set up non-interacting Green function g0(i\omega_n)
  S.G0_iw()[0](iw_) << 1.0 / (iw_ + mu + (-1.0) * delta_w(iw_));
  S.G0_iw()[1](iw_) << 1.0 / (iw_ + mu + (-1.0) * delta_w(iw_));

  // Solve!
  S.solve(ps);

  //if (mpi::communicator{}.rank() == 0) {
  //std::cout << "size= " << S.tau_samples[0](0, 0).size() << std::endl;
  //std::cout << "size= " << S.curlyG[0](0, 0).size() << std::endl;
  //for (auto [a, b] : zip(S.tau_samples[0](0, 0), S.weight_samples[0](0, 0))) { std::cout << "dtau= " << a << " weight= " << b << std::endl; }

  //auto archive = h5::file("anderson.out.h5", 'w');
  //h5_write(archive, "S", S);
  //}
}

File: test/python/static/anderson.py
from triqs_ctint import Solver

from itertools import product
import triqs.utility.mpi as mpi
from triqs.gf import *
from h5 import *
from triqs.operators import *
from triqs.utility.h5diff import h5diff

test_name = "anderson"

######## physical parameters ########
U = 1.0
mu = U/4.0
beta = 10.0
eps = 0.2

######## simulation parameters ########
n_cyc = 1000

# --------- set up static interactions and the block structure ---------
block_names = ['dn','up']
gf_struct = [(bl, 1) for bl in block_names]
h_int = U * n(block_names[0],0)*n(block_names[1],0)

# --------- Construct the ctint solver ----------
S = Solver(beta = beta,
               gf_struct = gf_struct,
               n_iw = 200,
               n_tau = 100001)

# --------- Initialize the non-interacting Green's function ----------
for bl, g_bl in S.G0_iw: g_bl << inverse(iOmega_n + mu - inverse(iOmega_n - eps));

# --------- Solve! ----------
S.solve(h_int=h_int,
        n_cycles = n_cyc,
        length_cycle = 50,
        n_warmup_cycles = 100,
        random_seed = 34788,
        measure_histogram = True,
        measure_density = True,
        measure_M4_iw = True,
        n_iw_M4 = 5,
        nfft_buf_size = 50,
        measure_M3pp_tau = True,
        measure_M3ph_tau = True,
        n_iw_M3 = 10,
        n_iW_M3 = 10,
        n_tau_M3 = 41,
        measure_chi2pp_tau = True,
        measure_chi2ph_tau = True,
        n_iw_chi2 = 10,
        n_tau_chi2 = 21,
        measure_chiAB_tau = True,
        chi_A_vec = [n('up',0) + n('dn', 0)],
        chi_B_vec = [n('up',0) + n('dn', 0)],
        post_process = True)

# -------- Save in archive ---------
with HDFArchive("%s.out.h5"%test_name,'w') as arch:
    arch["G0_iw"] = S.G0_iw
    arch["G_iw"] = S.G_iw
    arch["G2_iw"] = S.G2_iw
    arch["chi3pp_iw"] = S.chi3pp_iw
    arch["chi3ph_iw"] = S.chi3ph_iw
    arch["chi2pp_iw"] = S.chi2pp_iw
    arch["chi2ph_iw"] = S.chi2ph_iw
    arch["chiAB_iw"] = S.chiAB_iw
    arch["chi2pp_tau_from_M3"] = S.chi2pp_tau_from_M3
    arch["chi2ph_tau_from_M3"] = S.chi2ph_tau_from_M3

# -------- Compare ---------
h5diff("%s.out.h5"%test_name, "%s.ref.h5"%test_name)
